common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java

```diff
@@ -73,84 +77,145 @@
     private static final Object sLock = new Object();
     private static InteractiveTokenCommand sCommand = null;
     private static final CommandResultCache sCommandResultCache = new CommandResultCache();
-    private static final ConcurrentMap<BaseCommand, ResultFuture<CommandResult>> sExecutingCommandMap = new ConcurrentHashMap<>();
+
+    private static final Object mapAccessLock = new Object();
+    @GuardedBy("mapAccessLock")
+    private static ConcurrentMap<BaseCommand, FinalizableResultFuture<CommandResult>> sExecutingCommandMap = new ConcurrentHashMap<>();
+
+    /**
+     * Remove all keys that are the command reference from the executing command map.  Since if they key has
+     * been changed, remove will not work, construct a new map and add all keys that are not identically
+     * that key into the new map.
+     *
+     * @param command the command whose identity to use to cleanse the map.
+     */
+    private static void cleanMap(BaseCommand command) {
```

Comment:
synchronized?

Reply:
Maybe document that it must be used under the mapAccessLock?

Reply:
Can we wrap them here? synchronize block is reentrant right?

Reply:
I'm fine either way as soon as we can make sure the future dev are aware of this or can do the right thing)

Reply:
Yeah we could.  I'l like to dispose of this completely once we can stabilize the key space for this map either by generating stable keys or making the objects immutable.